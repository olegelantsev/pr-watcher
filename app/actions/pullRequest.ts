import fetch from 'cross-fetch';
import {
  GetState,
  Dispatch,
  PullRequest,
  User,
  Reviewer,
  AccountSetting,
  REQUEST_PULL_REQUESTS,
  RECEIVE_PULL_REQUESTS,
  ReceivePullRequestsAction,
  UPDATE_ACCOUNT,
  UpdateAccountSetting
} from '../reducers/types';

// https://docs.atlassian.com/bitbucket-server/rest/7.2.3/bitbucket-rest.html#idp281

export function updateAccount(
  accountSetting: AccountSetting
): UpdateAccountSetting {
  return {
    type: UPDATE_ACCOUNT,
    accountSetting
  };
}

export function requestPullRequests() {
  return {
    type: REQUEST_PULL_REQUESTS
  };
}

export async function fetchRepoSlags(baseUrl: string, projectKey: string) {
  const response = await fetch(
    `${baseUrl}/rest/api/1.0/projects/${projectKey}/repos`
  );
  const data = await response.json();
  return data.values.map((x: any) => x.slug);
}

function convertUser(x: any) {
  return {
    displayName: x.displayName,
    email: x.emailAddress,
    id: x.id,
    name: x.name
  } as User;
}

function parseReviewers(reviewers: [any]) {
  return reviewers.map((reviewer: any) => {
    return {
      approved: reviewer.approved,
      lastReviewedCommit: reviewer.lastReviewedCommit,
      user: convertUser(reviewer.user)
    } as Reviewer;
  });
}

function receivePosts(
  pullRequests: Array<PullRequest>
): ReceivePullRequestsAction {
  return {
    type: RECEIVE_PULL_REQUESTS,
    pullRequests,
    receivedAt: Date.now()
  };
}

export async function fetchPullsRequests(
  baseUrl: string,
  projectKey: string,
  repositorySlug: string
) {
  const response = await fetch(
    `${baseUrl}/rest/api/1.0/projects/${projectKey}/repos/${repositorySlug}/pull-requests`
  );
  const data = await response.json();
  return data.values.map((x: any) => {
    return {
      title: x.title,
      description: x.description,
      createdDate: x.createdDate,
      updatedDate: x.updatedDate,
      author: convertUser(x.author.user),
      reviewers: parseReviewers(x.reviewers)
    } as PullRequest;
  });
}

async function fetchPullRequestsAndMerge(
  repoSlugs: Array<string>,
  accountSetting: AccountSetting
): Promise<Array<PullRequest>> {
  const listOfPullRequests = await Promise.all(
    repoSlugs.map((repoSlug: string) =>
      fetchPullsRequests(
        accountSetting.url,
        accountSetting.projectSlug,
        repoSlug
      )
    )
  );
  const merged = [].concat(...listOfPullRequests);

  return merged;
}

export function fetchAll() {
  return (dispatch: Dispatch, getState: GetState) => {
    const { accountSetting } = getState();

    if (accountSetting == null) {
      throw Error("Can't fetch if account isn't added.");
    }

    dispatch(requestPullRequests());

    return fetchRepoSlags(accountSetting.url, accountSetting.projectSlug)
      .then(repoSlugs => {
        return fetchPullRequestsAndMerge(repoSlugs, accountSetting);
      })
      .then((pullRequests: Array<PullRequest>) => {
        return dispatch(receivePosts(pullRequests));
      })
      .catch(err => {
        throw err;
      });
  };
}
