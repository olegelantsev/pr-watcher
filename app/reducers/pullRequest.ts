import {
  RECEIVE_PULL_REQUESTS,
  REQUEST_PULL_REQUESTS,
  UPDATE_ACCOUNT,
  PullRequestActions,
  PullRequests
} from './types';

const initialState = {
  fetchInProgress: false,
  newPullRequests: 0,
  prs: []
} as PullRequests;

export default function pullRequests(
  state: PullRequests = initialState,
  action: PullRequestActions
): PullRequests {
  switch (action.type) {
    case RECEIVE_PULL_REQUESTS:
      return {
        ...state,
        fetchInProgress: false
      };
    case REQUEST_PULL_REQUESTS:
      return {
        ...state,
        fetchInProgress: true
      };
    case UPDATE_ACCOUNT:
      return {
        ...state
      };
    default:
      return state;
  }
}
