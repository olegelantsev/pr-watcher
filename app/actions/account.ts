import {
  AccountCheckError,
  AccountSetting,
  UPDATE_ACCOUNT,
  UpdateAccountSetting,
  errorType,
  AccountCheckProgress,
  ACCOUNT_CHECK_IN_PROGRESS,
  Dispatch,
  ACCOUNT_CHECK_ERROR
} from '../reducers/types';
import { fetchRepoSlags } from './pullRequest';

export function updateAccount(
  accountSetting: AccountSetting,
  checking: boolean,
  error?: errorType
): UpdateAccountSetting {
  return {
    type: UPDATE_ACCOUNT,
    accountSetting,
    checking,
    error
  };
}

function accountCheckInProgress(): AccountCheckProgress {
  return {
    type: ACCOUNT_CHECK_IN_PROGRESS
  };
}

function accountCheckError(
  status: errorType,
  error: string
): AccountCheckError {
  return {
    type: ACCOUNT_CHECK_ERROR,
    error,
    status
  };
}

async function checkAccount(accountSetting: AccountSetting): Promise<boolean> {
  await fetchRepoSlags(accountSetting);
  return true;
}

export function checkAndUpdateAccount(accountSetting: AccountSetting) {
  return (dispatch: Dispatch) => {
    dispatch(accountCheckInProgress());

    return checkAccount(accountSetting)
      .then(() => {
        return dispatch(updateAccount(accountSetting, false, undefined));
      })
      .catch(err => {
        dispatch(accountCheckError('error', String(err)));
      });
  };
}
