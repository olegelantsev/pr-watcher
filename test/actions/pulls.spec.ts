import { spy } from 'sinon';
import nock from 'nock';
import * as actions from '../../app/actions/pullRequest';
import * as repoResponse from './stubs/repos.json';
import * as pullRequestsResponse from './stubs/pullRequests.json';
import * as twoReposResponse from './stubs/repos2.json';
import * as repo2PullRequestResponse from './stubs/repo2_pullRequests.json';
import {
  REQUEST_PULL_REQUESTS,
  ApplicationState,
  AccountSetting
} from '../../app/reducers/types';

describe('actions', () => {
  it('should fetch list of repo slugs', async () => {
    nock('http://bitbucket.com')
      .get('/rest/api/1.0/projects/123/repos')
      .reply(200, repoResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      });
    const repoSlugs = await actions.fetchRepoSlags({
      url: 'http://bitbucket.com',
      token: 'some-token',
      projectSlug: '123'
    } as AccountSetting);
    expect(repoSlugs.sort()).toEqual(['repo1']);
  });

  it('should fetch list of PRs in repo', async () => {
    nock('http://bitbucket.com')
      .get('/rest/api/1.0/projects/projectname/repos/reponame/pull-requests')
      .reply(200, pullRequestsResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      });
    const pullRequests = await actions.fetchPullsRequests(
      'http://bitbucket.com',
      'projectname',
      'reponame'
    );
    expect(pullRequests).toMatchSnapshot();
  });

  it('should fetch repos and pull requests', () => {
    nock('http://bitbucket.com')
      .get('/rest/api/1.0/projects/projectname/repos')
      .reply(200, twoReposResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      })
      .get('/rest/api/1.0/projects/projectname/repos/repo1/pull-requests')
      .reply(200, pullRequestsResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      })
      .get('/rest/api/1.0/projects/projectname/repos/repo2/pull-requests')
      .reply(200, repo2PullRequestResponse, {
        'Access-Control-Allow-Origin': '*',
        'Content-type': 'application/json'
      });
    return new Promise(resolve => {
      Date.now = jest.fn(() => 1359075920);
      const fn = actions.fetchAll();
      const dispatch = spy();
      const getState = () =>
        ({
          accountSetting: {
            projectSlug: 'projectname',
            token: 'mytoken',
            url: 'http://bitbucket.com'
          }
        } as ApplicationState);
      fn(dispatch, getState);
      setTimeout(() => {
        const calls = dispatch.getCalls();
        expect(calls.length).toBe(2);

        expect(dispatch.getCall(0).args[0]).toEqual({
          type: REQUEST_PULL_REQUESTS
        });

        expect(dispatch.getCall(1).args[0]).toMatchSnapshot();
        resolve();
      }, 40);
    });
  });

  it('should dispatch error if fetch fails', () => {
    return new Promise(resolve => {
      Date.now = jest.fn(() => 1359075920);
      const fn = actions.fetchAll();
      const dispatch = spy();
      const getState = () =>
        ({
          accountSetting: {
            projectSlug: 'projectname',
            token: 'mytoken',
            url: 'http://example.com'
          }
        } as ApplicationState);
      fn(dispatch, getState);
      setTimeout(() => {
        const calls = dispatch.getCalls();
        expect(calls.length).toBe(2);

        expect(dispatch.getCall(1).args[0]).toMatchSnapshot();
        resolve();
      }, 1500);
    });
  });
});
