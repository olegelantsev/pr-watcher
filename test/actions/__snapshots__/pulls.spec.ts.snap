// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`actions should fetch list of PRs in repo 1`] = `
Array [
  Object {
    "author": Object {
      "displayName": "Tom",
      "email": "tom@example.com",
      "id": 115026,
      "name": "tom",
    },
    "createdDate": 1359075920,
    "description": "It’s a kludge, but put the tuple from the database in the cache.",
    "reviewers": Array [
      Object {
        "approved": true,
        "lastReviewedCommit": "7549846524f8aed2bd1c0249993ae1bf9d3c9998",
        "user": Object {
          "displayName": "Jane Citizen",
          "email": "jane@example.com",
          "id": 101,
          "name": "jcitizen",
        },
      },
    ],
    "title": "Talking Nerdy",
    "updatedDate": 1359085920,
  },
]
`;

exports[`actions should fetch repos and pull requests 1`] = `
Object {
  "pullRequests": Array [
    Object {
      "author": Object {
        "displayName": "Tom",
        "email": "tom@example.com",
        "id": 115026,
        "name": "tom",
      },
      "createdDate": 1359075920,
      "description": "It’s a kludge, but put the tuple from the database in the cache.",
      "reviewers": Array [
        Object {
          "approved": true,
          "lastReviewedCommit": "7549846524f8aed2bd1c0249993ae1bf9d3c9998",
          "user": Object {
            "displayName": "Jane Citizen",
            "email": "jane@example.com",
            "id": 101,
            "name": "jcitizen",
          },
        },
      ],
      "title": "Talking Nerdy",
      "updatedDate": 1359085920,
    },
    Object {
      "author": Object {
        "displayName": "Tom",
        "email": "tom@example.com",
        "id": 115026,
        "name": "tom",
      },
      "createdDate": 1359075920,
      "description": "It’s a kludge, but put the tuple from the database in the cache.",
      "reviewers": Array [
        Object {
          "approved": true,
          "lastReviewedCommit": "7549846524f8aed2bd1c0249993ae1bf9d3c9998",
          "user": Object {
            "displayName": "Jane Citizen",
            "email": "jane@example.com",
            "id": 101,
            "name": "jcitizen",
          },
        },
      ],
      "title": "Second PR",
      "updatedDate": 1359085920,
    },
  ],
  "receivedAt": 1359075920,
  "type": "RECEIVE_PULL_REQUESTS",
}
`;
